File: src/main/java/org/samoxive/safetyjim/discord/DiscordShard.java
package org.samoxive.safetyjim.discord;

import net.dv8tion.jda.core.AccountType;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.JDABuilder;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.*;
import net.dv8tion.jda.core.events.guild.GuildJoinEvent;
import net.dv8tion.jda.core.events.guild.GuildLeaveEvent;
import net.dv8tion.jda.core.events.guild.member.GuildMemberJoinEvent;
import net.dv8tion.jda.core.events.guild.member.GuildMemberLeaveEvent;
import net.dv8tion.jda.core.events.message.guild.GuildMessageDeleteEvent;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.core.events.message.guild.react.GuildMessageReactionAddEvent;
import net.dv8tion.jda.core.events.message.guild.react.GuildMessageReactionRemoveEvent;
import net.dv8tion.jda.core.exceptions.RateLimitedException;
import net.dv8tion.jda.core.hooks.ListenerAdapter;
import net.dv8tion.jda.core.managers.GuildController;
import net.dv8tion.jda.core.requests.SessionReconnectQueue;
import org.jooq.DSLContext;
import org.jooq.Result;
import org.samoxive.jooq.generated.Tables;
import org.samoxive.jooq.generated.tables.records.*;
import org.samoxive.safetyjim.config.Config;
import org.samoxive.safetyjim.database.DatabaseUtils;
import org.samoxive.safetyjim.discord.commands.Mute;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.security.auth.login.LoginException;
import java.awt.*;
import java.sql.Timestamp;
import java.util.*;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.stream.Collectors;

public class DiscordShard extends ListenerAdapter {
    private Logger log;
    private DiscordBot bot;
    private JDA shard;
    private ExecutorService threadPool;

    public DiscordShard(DiscordBot bot, int shardId) {
        this.bot = bot;
        log = LoggerFactory.getLogger("DiscordShard " + DiscordUtils.getShardString(shardId, bot.getConfig().jim.shard_count));

        Config config = bot.getConfig();
        int shardCount = config.jim.shard_count;
        String version = config.version;

        threadPool = Executors.newCachedThreadPool();
        JDABuilder builder = new JDABuilder(AccountType.BOT);
        try {
            this.shard = builder.setToken(bot.getConfig().jim.token)
                                .setAudioEnabled(false) // jim doesn't have any audio functionality
                                .addEventListener(this)
                                .setReconnectQueue(new SessionReconnectQueue()) // needed to prevent shards trying to reconnect too soon
                                .setEnableShutdownHook(true)
                                .useSharding(shardId, bot.getConfig().jim.shard_count)
                                .setGame(Game.playing(String.format("-mod help | %s | %s", version, DiscordUtils.getShardString(shardId, shardCount))))
                                .buildBlocking();
        } catch (LoginException e) {
            log.error("Invalid token.");
            System.exit(1);
        } catch (InterruptedException e) {
            log.error("Something something", e);
            System.exit(1);
        } catch (RateLimitedException e) {
            log.error("Hit Discord API Rate Limit", e);
            System.exit(1);
        }
    }

    private void populateStatistics(JDA shard) {
        DSLContext database = bot.getDatabase();
        shard.getGuilds()
             .stream()
             .filter((guild) -> DatabaseUtils.getGuildSettings(database, guild).getStatistics())
             .forEach((guild -> populateGuildStatistics(guild)));
    }

    public void populateGuildStatistics(Guild guild) {
        DSLContext database = bot.getDatabase();
        List<TextChannel> channels = guild.getTextChannels();
        for(TextChannel channel: channels) {
            MessagesRecord oldestRecord = database.selectFrom(Tables.MESSAGES)
                    .where(Tables.MESSAGES.GUILDID.eq(guild.getId()))
                    .and(Tables.MESSAGES.CHANNELID.eq(channel.getId()))
                    .orderBy(Tables.MESSAGES.DATE.asc())
                    .limit(1)
                    .fetchAny();

            MessagesRecord newestRecord = database.selectFrom(Tables.MESSAGES)
                    .where(Tables.MESSAGES.GUILDID.eq(guild.getId()))
                    .and(Tables.MESSAGES.CHANNELID.eq(channel.getId()))
                    .orderBy(Tables.MESSAGES.DATE.desc())
                    .limit(1)
                    .fetchAny();

            List<Message> fetchedMessages = null;
            if (oldestRecord == null || newestRecord == null) {
                fetchedMessages = DiscordUtils.fetchHistoryFromScratch(channel);
            } else {
                Message oldestMessageStored = null, newestMessageStored = null;

                try {
                    oldestMessageStored = channel.getMessageById(oldestRecord.getMessageid()).complete();
                    newestMessageStored = channel.getMessageById(newestRecord.getMessageid()).complete();
                    if (oldestMessageStored == null || newestMessageStored == null) {
                        throw new Exception();
                    }

                    fetchedMessages = DiscordUtils.fetchFullHistoryBeforeMessage(channel, oldestMessageStored);
                    fetchedMessages.addAll(DiscordUtils.fetchFullHistoryAfterMessage(channel, newestMessageStored));
                } catch (Exception e) {
                    database.deleteFrom(Tables.MESSAGES)
                            .where(Tables.MESSAGES.CHANNELID.eq(channel.getId()))
                            .and(Tables.MESSAGES.GUILDID.eq(guild.getId()))
                            .execute();
                    fetchedMessages = DiscordUtils.fetchHistoryFromScratch(channel);
                }
            }

            if (fetchedMessages.size() == 0) {
                continue;
            }

            List<MessagesRecord> records = fetchedMessages.stream()
                    .map(message -> {
                        MessagesRecord record = database.newRecord(Tables.MESSAGES);
                        User user = message.getAuthor();
                        String content = message.getRawContent();
                        int wordCount = content.split(" ").length;
                        record.setMessageid(message.getId());
                        record.setUserid(user.getId());
                        record.setChannelid(channel.getId());
                        record.setGuildid(channel.getGuild().getId());
                        record.setDate(DiscordUtils.getCreationTime(message.getId()));
                        record.setWordcount(wordCount);
                        record.setSize(content.length());
                        return record;
                    })
                    .collect(Collectors.toList());

            database.batchStore(records).execute();
        }
    }

    @Override
    public void onReady(ReadyEvent event) {
        log.info("Shard is ready.");
        DSLContext database = this.bot.getDatabase();
        JDA shard = event.getJDA();

        for (Guild guild: shard.getGuilds()) {
            if (DiscordUtils.isBotFarm(guild)) {
                guild.leave().queue();
            }
        }

        int guildsWithMissingKeys = 0;
        for (Guild guild: shard.getGuilds()) {
            SettingsRecord guildSettings = DatabaseUtils.getGuildSettings(database, guild);

            if (guildSettings == null) {
                DatabaseUtils.deleteGuildSettings(database, guild);
                DatabaseUtils.createGuildSettings(this.bot, database, guild);
                guildsWithMissingKeys++;
            }
        }

        if (guildsWithMissingKeys > 0) {
            log.warn("Added {} guild(s) to the database with invalid number of settings.", guildsWithMissingKeys);
        }

        threadPool.submit(() -> {
            try {
                populateStatistics(shard);
            } catch (Exception e) {
                e.printStackTrace();
            }
            log.info("Populated statistics.");
        });
    }

    @Override
    public void onGuildMessageReceived(GuildMessageReceivedEvent event) {
        if (event.getAuthor().isBot()) {
            return;
        }

        DSLContext database = bot.getDatabase();
        Guild guild = event.getGuild();
        Message message = event.getMessage();
        String content = message.getRawContent();
        JDA shard = event.getJDA();
        SelfUser self = shard.getSelfUser();

        if (message.isMentioned(self) && content.contains("prefix")) {
            SettingsRecord guildSettings = DatabaseUtils.getGuildSettings(database, guild);
            String prefix = guildSettings.getPrefix();
            DiscordUtils.successReact(bot, message);

            EmbedBuilder embed = new EmbedBuilder();
            embed.setAuthor("Safety Jim - Prefix", null, self.getAvatarUrl())
                 .setDescription("This guild's prefix is: " + prefix)
                 .setColor(new Color(0x4286F4));

            DiscordUtils.sendMessage(message.getTextChannel(), embed.build());
            return;
        }

        List<Future<Boolean>> processorResults = new LinkedList<>();

        // Spread processing jobs across threads as they are likely to be independent of io operations
        for (MessageProcessor processor: bot.getProcessors()) {
            Future<Boolean> future = threadPool.submit(() -> processor.onMessage(bot, this, event));
            processorResults.add(future);
        }

        // If processors return true, that means they deleted the original message so we don't need to continue further
        for (Future<Boolean> result: processorResults) {
            try {
                if (result.get().equals(true)) {
                    return;
                }
            } catch (Exception e) {
                //
            }
        }

        SettingsRecord guildSettings = DatabaseUtils.getGuildSettings(database, guild);
        if (guildSettings == null) { // settings aren't initialized yet
            return;
        }
        String prefix = guildSettings.getPrefix().toLowerCase();

        // 0 = prefix, 1 = command, rest are accepted as arguments
        String[] splitContent = content.trim().split(" ");
        String firstWord = splitContent[0].toLowerCase();
        Command command;
        String commandName;

        if (!guildSettings.getNospaceprefix()) {
            if (!firstWord.equals(prefix)) {
                return;
            }

            // This means the user only entered the prefix
            if (splitContent.length == 1) {
                DiscordUtils.failReact(bot, message);
                return;
            }

            // We also want commands to be case insensitive
            commandName = splitContent[1].toLowerCase();
            command = bot.getCommands().get(commandName);
        } else {
            if (!firstWord.startsWith(prefix)) {
                return;
            }

            if (firstWord.length() == prefix.length()) {
                DiscordUtils.failReact(bot, message);
                return;
            }

            commandName = firstWord.substring(prefix.length());
            command = bot.getCommands().get(commandName);
        }

        // Command not found
        if (command == null) {
            DiscordUtils.failReact(bot, message);
            return;
        }

        // Join words back with whitespace as some commands don't need them split,
        // they can split the arguments again if needed
        StringJoiner args = new StringJoiner(" ");
        int startIndex = guildSettings.getNospaceprefix() ? 1 : 2;
        for (int i = startIndex; i < splitContent.length; i++) {
            args.add(splitContent[i]);
        }

        // Command executions are likely to be io dependant, better send them in a seperate thread to not block
        // discord client
        threadPool.execute(() -> executeCommand(event, command, commandName, args.toString().trim()));
    }

    @Override
    public void onException(ExceptionEvent event) {
        log.error("An exception occurred.", event.getCause());
    }

    @Override
    public void onGuildMessageDelete(GuildMessageDeleteEvent event) {
        // TODO(sam): Add message cache and trigger message processors if
        // deleted message is in the cache
    }

    @Override
    public void onGuildMessageReactionAdd(GuildMessageReactionAddEvent event) {
        if (event.getMember().getUser().isBot() || event.getChannelType() != ChannelType.TEXT) {
            return;
        }

        for (MessageProcessor processor: bot.getProcessors()) {
            threadPool.execute(() -> processor.onReactionAdd(bot, this, event));
        }
    }

    @Override
    public void onGuildMessageReactionRemove(GuildMessageReactionRemoveEvent event) {
        if (event.getMember().getUser().isBot() || event.getChannelType() != ChannelType.TEXT) {
            return;
        }

        for (MessageProcessor processor: bot.getProcessors()) {
            threadPool.execute(() -> processor.onReactionRemove(bot, this, event));
        }
    }

    @Override
    public void onGuildJoin(GuildJoinEvent event) {
        if (DiscordUtils.isBotFarm(event.getGuild())) {
            event.getGuild().leave().queue();
            return;
        }

        Guild guild = event.getGuild();
        DSLContext database = bot.getDatabase();
        String message = String.format("Hello! I am Safety Jim, `%s` is my default prefix!", bot.getConfig().jim.default_prefix);
        DiscordUtils.sendMessage(DiscordUtils.getDefaultChannel(guild), message);
        DatabaseUtils.createGuildSettings(bot, database, guild);
    }

    @Override
    public void onGuildLeave(GuildLeaveEvent event) {
        DatabaseUtils.deleteGuildSettings(bot.getDatabase(), event.getGuild());
    }

    @Override
    public void onGuildMemberJoin(GuildMemberJoinEvent event) {
        JDA shard = event.getJDA();
        Guild guild = event.getGuild();
        GuildController controller = guild.getController();
        Member member = event.getMember();
        User user = member.getUser();
        DSLContext database = bot.getDatabase();
        SettingsRecord guildSettings = DatabaseUtils.getGuildSettings(database, guild);

        if (guildSettings.getWelcomemessage()) {
            String textChannelId = guildSettings.getWelcomemessagechannelid();
            TextChannel channel = shard.getTextChannelById(textChannelId);
            if (channel != null) {
                String message = guildSettings.getMessage()
                                         .replace("$user", member.getAsMention())
                                         .replace("$guild", guild.getName());
                if (guildSettings.getHoldingroom()) {
                    String waitTime = guildSettings.getHoldingroomminutes().toString();
                    message = message.replace("$minute", waitTime);
                }

                DiscordUtils.sendMessage(channel, message);
            }
        }

        if (guildSettings.getHoldingroom()) {
            int waitTime = guildSettings.getHoldingroomminutes();
            long currentTime = System.currentTimeMillis() / 1000;

            JoinlistRecord newRecord = database.newRecord(Tables.JOINLIST);
            newRecord.setUserid(member.getUser().getId());
            newRecord.setGuildid(guild.getId());
            newRecord.setJointime(currentTime);
            newRecord.setAllowtime(currentTime + waitTime * 60);
            newRecord.setAllowed(false);
            newRecord.store();
        }


        Result<MutelistRecord> records = database.selectFrom(Tables.MUTELIST)
                                               .where(Tables.MUTELIST.GUILDID.eq(guild.getId()))
                                               .and(Tables.MUTELIST.USERID.eq(user.getId()))
                                               .fetch();

        if (records.isEmpty()) {
            return;
        }

        Role mutedRole = null;
        try {
            mutedRole = Mute.setupMutedRole(guild);
        } catch (Exception e) {
            return;
        }

        try {
            controller.addSingleRoleToMember(member, mutedRole).complete();
        } catch (Exception e) {
            // Maybe actually do something if this fails?
        }
    }

    @Override
    public void onGuildMemberLeave(GuildMemberLeaveEvent event) {
        bot.getDatabase()
           .deleteFrom(Tables.JOINLIST)
           .where(Tables.JOINLIST.USERID.eq(event.getUser().getId()))
           .execute();
    }

    public JDA getShard() {
        return shard;
    }

    private void createCommandLog(GuildMessageReceivedEvent event, String commandName, String args, Date time, long from, long to) {
        User author = event.getAuthor();
        CommandlogsRecord record = bot.getDatabase().newRecord(Tables.COMMANDLOGS);
        record.setCommand(commandName);
        record.setArguments(args);
        record.setTime(new Timestamp(time.getTime()));
        record.setUsername(DiscordUtils.getTag(author));
        record.setUserid(author.getId());
        record.setGuildname(event.getGuild().getName());
        record.setGuildid(event.getGuild().getId());
        record.setExecutiontime((int) (to - from));
        record.store();
    }

    private void executeCommand(GuildMessageReceivedEvent event, Command command, String commandName, String args) {
        JDA shard = event.getJDA();

        Date date = new Date();
        long startTime = System.currentTimeMillis();
        boolean showUsage = false;
        try {
            showUsage = command.run(bot, event, args);
        } catch (Exception e) {
            DiscordUtils.failReact(bot, event.getMessage());
            DiscordUtils.sendMessage(event.getChannel(), "There was an error running your command, this incident has been logged.");
            log.error(String.format("%s failed with arguments %s in guild %s - %s", commandName, args, event.getGuild().getName(), event.getGuild().getId()), e);
        } finally {
            long endTime = System.currentTimeMillis();
            threadPool.submit(() -> createCommandLog(event, commandName, args, date, startTime, endTime));
        }

        if (showUsage) {
            String[] usages = command.getUsages();
            SettingsRecord guildSettings = DatabaseUtils.getGuildSettings(bot.getDatabase(), event.getGuild());
            String prefix = guildSettings.getPrefix();

            EmbedBuilder embed = new EmbedBuilder();
            embed.setAuthor("Safety Jim - \"" + commandName + "\" Syntax", null, shard.getSelfUser().getAvatarUrl())
                 .setDescription(DiscordUtils.getUsageString(prefix, usages))
                 .setColor(new Color(0x4286F4));

            DiscordUtils.failReact(bot, event.getMessage());
            event.getChannel().sendMessage(embed.build()).queue();
        } else {
            String[] deleteCommands = {
                    "ban", "kick", "mute", "softban", "warn"
            };

            for (String deleteCommand: deleteCommands) {
                if (commandName.equals(deleteCommand)) {
                    DiscordUtils.deleteCommandMessage(bot, event.getMessage());
                    return;
                }
            }
        }
    }

    public ExecutorService getThreadPool() {
        return threadPool;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/DiscordShard.kt
package org.samoxive.safetyjim.discord

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.entities.Activity
import net.dv8tion.jda.api.entities.Role
import net.dv8tion.jda.api.events.ExceptionEvent
import net.dv8tion.jda.api.events.ReadyEvent
import net.dv8tion.jda.api.events.guild.GuildJoinEvent
import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent
import net.dv8tion.jda.api.events.guild.member.GuildMemberRemoveEvent
import net.dv8tion.jda.api.events.message.guild.GuildMessageDeleteEvent
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.utils.ChunkingFilter
import net.dv8tion.jda.api.utils.MemberCachePolicy
import net.dv8tion.jda.api.utils.SessionController
import net.dv8tion.jda.api.utils.cache.CacheFlag
import org.samoxive.safetyjim.database.*
import org.samoxive.safetyjim.database.SettingsTable.getGuildSettings
import org.samoxive.safetyjim.discord.commands.setupMutedRole
import org.samoxive.safetyjim.discord.processors.isInviteLinkBlocklisted
import org.samoxive.safetyjim.tryhardAsync
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.Color
import java.util.*
import javax.security.auth.login.LoginException
import kotlin.system.exitProcess

const val DEPRECATION_NOTICE = "Due to Discord's changes, the old way of implementing commands as text messages has been deprecated.\n" +
    "Safety Jim will migrate to slash commands on 8th of January after much work rewriting most of the project to adapt to changes.\n" +
    "For this migration there is no action needed, however if you invited Jim after 24th of March, 2021 you need to kick him and invite back for slash commands to appear in your server.\n" +
    "You can use the whois command with Jim to find when you invited him.\n\n" +
    "To check out the new slash commands before migration and report potential bugs and feedback, feel free to head to the support Discord server."

class DiscordShard(private val bot: DiscordBot, shardId: Int, sessionController: SessionController) : ListenerAdapter() {
    private val log: Logger
    val jda: JDA
    val confirmationListener = ConfirmationListener()

    init {
        val config = bot.config
        val shardString = getShardString(shardId, config.jim.shard_count)
        log = LoggerFactory.getLogger("DiscordShard $shardString")

        this.jda = try {
            JDABuilder.create(
                config.jim.token,
                GatewayIntent.GUILD_MEMBERS,
                GatewayIntent.GUILD_MESSAGES
            )
                .setChunkingFilter(ChunkingFilter.NONE)
                .addEventListeners(this, confirmationListener)
                .setSessionController(sessionController) // needed to prevent shards trying to reconnect too soon
                .useSharding(shardId, config.jim.shard_count)
                .disableCache(
                    EnumSet.of(
                        CacheFlag.ACTIVITY,
                        CacheFlag.VOICE_STATE,
                        CacheFlag.EMOTE,
                        CacheFlag.CLIENT_STATUS,
                        CacheFlag.ONLINE_STATUS
                    )
                )
                .setMemberCachePolicy(MemberCachePolicy.NONE)
                .setActivity(Activity.playing("safetyjim.xyz $shardString"))
                .build()
                .awaitReady()
        } catch (e: LoginException) {
            log.error("Invalid token.", e)
            exitProcess(1)
        } catch (e: InterruptedException) {
            log.error("Something something", e)
            exitProcess(1)
        }
    }

    override fun onReady(event: ReadyEvent) {
        log.info("Shard is ready.")
    }

    override fun onGuildMessageReceived(event: GuildMessageReceivedEvent) {
        GlobalScope.launch { onGuildMessageReceivedAsync(event) }
    }

    private suspend fun onGuildMessageReceivedAsync(event: GuildMessageReceivedEvent) {
        if (event.author.isBot) {
            return
        }

        // make sure event#member is not null, it's used quite a lot
        if (event.isWebhookMessage) {
            return
        }

        val guild = event.guild
        val message = event.message
        val content = message.contentRaw
        val shard = event.jda
        val self = shard.selfUser

        if (message.isMentioned(self) && content.contains("prefix")) {
            val guildSettings = getGuildSettings(guild, bot.config)
            val prefix = guildSettings.prefix
            message.successReact()

            val embed = EmbedBuilder()
            embed.setAuthor("Safety Jim - Prefix", null, self.avatarUrl)
                .setDescription("This guild's prefix is: $prefix")
                .setColor(Color(0x4286F4))

            message.textChannel.trySendMessage(embed.build(), message)
            return
        }

        val guildSettings = getGuildSettings(guild, bot.config)

        // Spread processing jobs across threads as they are likely to be independent of io operations
        val processorResult = bot.processors.map {
            GlobalScope.async {
                tryhardAsync {
                    it.onMessage(bot, this@DiscordShard, event, guildSettings)
                } ?: false
            }
        }.map { it.await() }.reduce { acc, elem -> acc || elem }

        // If processors return true, that means they deleted the original message, so we don't need to continue further
        if (processorResult) {
            return
        }

        val prefix = guildSettings.prefix.lowercase()

        // 0 = prefix, 1 = command, rest are accepted as arguments
        val splitContent = content.trim().split(" ").toTypedArray()
        val firstWord = splitContent[0].lowercase()
        val command: Command?
        val commandName: String

        if (!guildSettings.noSpacePrefix) {
            if (firstWord != prefix) {
                return
            }

            // This means the user only entered the prefix
            if (splitContent.size == 1) {
                message.failReact()
                return
            }

            // We also want commands to be case-insensitive
            commandName = splitContent[1].lowercase()
            command = bot.commands[commandName]
        } else {
            if (!firstWord.startsWith(prefix)) {
                return
            }

            if (firstWord.length == prefix.length) {
                message.failReact()
                return
            }

            commandName = firstWord.substring(prefix.length)
            command = bot.commands[commandName]
        }

        // Command not found
        if (command == null) {
            message.failReact()
            return
        }

        // Join words back with whitespace as some commands don't need them split,
        // they can split the arguments again if needed
        val args = StringJoiner(" ")
        val startIndex = if (guildSettings.noSpacePrefix) 1 else 2
        for (i in startIndex until splitContent.size) {
            args.add(splitContent[i])
        }

        // Command executions are likely to be io dependant, better send them in a seperate thread to not block
        // discord client
        executeCommand(event, guildSettings, command, commandName, args.toString().trim())
    }

    override fun onGuildMessageDelete(event: GuildMessageDeleteEvent) {
        bot.processors.forEach { processor -> GlobalScope.launch { processor.onMessageDelete(bot, this@DiscordShard, event) } }
    }

    override fun onException(event: ExceptionEvent) {
        log.error("An exception occurred.", event.cause)
    }

    override fun onGuildJoin(event: GuildJoinEvent) {
        GlobalScope.launch {
            val guild = event.guild

            val defaultPrefix = bot.config.jim.default_prefix
            val message = "Hello! I am Safety Jim, `$defaultPrefix` is my default prefix! Visit https://safetyjim.xyz/commands to see available commands.\nYou can join the support server at https://discord.io/safetyjim or contact Samoxive#8634 for help."
            guild.getDefaultChannelTalkable()?.trySendMessage(message)
            SettingsTable.insertDefaultGuildSettings(bot.config, guild)
        }
    }

    override fun onGuildMemberJoin(event: GuildMemberJoinEvent) {
        GlobalScope.launch { onGuildMemberJoinAsync(event) }
    }

    private suspend fun onGuildMemberJoinAsync(event: GuildMemberJoinEvent) {
        val shard = event.jda
        val guild = event.guild
        val member = event.member
        val user = member.user
        val guildSettings = getGuildSettings(guild, bot.config)

        if (guildSettings.inviteLinkRemover) {
            if (isInviteLinkBlocklisted(user.name)) {
                tryhardAsync { guild.kick(member).await() }
                return
            }
        }

        if (guildSettings.welcomeMessage) {
            val textChannelId = guildSettings.welcomeMessageChannelId
            val channel = shard.getTextChannelById(textChannelId)
            if (channel != null) {
                var message = guildSettings.message
                    .replace("\$user", member.asMention)
                    .replace("\$guild", guild.name)
                if (guildSettings.holdingRoom) {
                    val waitTime = guildSettings.holdingRoomMinutes.toString()
                    message = message.replace("\$minute", waitTime)
                }

                channel.trySendMessage(message)
            }
        }

        if (guildSettings.holdingRoom) {
            val waitTime = guildSettings.holdingRoomMinutes
            val currentTime = System.currentTimeMillis() / 1000

            JoinsTable.insertJoin(
                JoinEntity(
                    userId = member.user.idLong,
                    guildId = guild.idLong,
                    joinTime = currentTime,
                    allowTime = currentTime + waitTime * 60,
                    allowed = false
                )
            )
        }

        if (guildSettings.joinCaptcha) {
            val captchaUrl = "${bot.config.server.self_url}captcha/${guild.id}/${user.id}"
            user.trySendMessage("Welcome to ${guild.name}! To enter you must complete this captcha.\n$captchaUrl")
        }

        val records = MutesTable.fetchValidUserMutes(guild, user)

        if (records.isEmpty()) {
            return
        }

        val mutedRole: Role = tryhardAsync { setupMutedRole(guild) } ?: return
        tryhardAsync { guild.addRoleToMember(member, mutedRole).await() }
    }

    override fun onGuildMemberRemove(event: GuildMemberRemoveEvent) {
        GlobalScope.launch {
            JoinsTable.deleteUserJoins(event.guild, event.user)
        }
    }

    private suspend fun executeCommand(event: GuildMessageReceivedEvent, settings: SettingsEntity, command: Command, commandName: String, args: String) {
        val shard = event.jda
        val message = event.message
        val channel = event.channel

        var showUsage = false
        try {
            showUsage = command.run(bot, event, settings, args)
        } catch (e: Exception) {
            message.failReact()
            channel.trySendMessage("There was an error running your command, this incident has been logged.", message)
            log.error("$commandName failed with arguments $args in guild ${event.guild.name} - ${event.guild.id}", e)
        }

        if (showUsage) {
            val usages = command.usages
            val guildSettings = getGuildSettings(event.guild, bot.config)
            val prefix = guildSettings.prefix

            val embed = EmbedBuilder()
            embed.setAuthor("Safety Jim - \"$commandName\" Syntax", null, shard.selfUser.avatarUrl)
                .setDescription(getUsageString(prefix, usages))
                .setColor(Color(0x4286F4))

            message.failReact()
            channel.trySendMessage(embed.build(), message)
        } else {
            message.deleteCommandMessage(settings, commandName)
        }

        val guildId = event.guild.idLong
        if (!bot.notifiedGuilds.contains(guildId)) {
            bot.notifiedGuilds.add(guildId)
            channel.trySendMessage(DEPRECATION_NOTICE)
        }
    }
}
